pkg/extract/compression.go

```diff
@@ -0,0 +1,137 @@
+package extract
+
+import (
+	"compress/bzip2"
+	"compress/gzip"
+	"compress/lzw"
+	"encoding/binary"
+	"io"
+
+	"github.com/pierrec/lz4"
+	"github.com/ulikunitz/xz"
+
+	"github.com/replicate/pget/pkg/logging"
+)
+
+const (
+	peekSize = 8
+
+	gzipMagic = 0x1F8B
+	bzipMagic = 0x425A
+	xzMagic   = 0xFD377A585A00
+	lzwMagic  = 0x1F9D
+	lz4Magic  = 0x184D2204
+)
+
+var _ decompressor = gzipDecompressor{}
+var _ decompressor = bzip2Decompressor{}
+var _ decompressor = xzDecompressor{}
+var _ decompressor = lzwDecompressor{}
+var _ decompressor = lz4Decompressor{}
+var _ decompressor = noOpDecompressor{}
+
+// decompressor represents different compression formats.
+type decompressor interface {
+	decompress(r io.Reader) (io.Reader, error)
+}
+
+// detectFormat returns the appropriate extractor according to the magic number.
+func detectFormat(input []byte) decompressor {
+	log := logging.GetLogger()
+	inputSize := len(input)
+
+	if inputSize < 2 {
+		return noOpDecompressor{}
+	}
+	// pad to 8 bytes
+	if inputSize < 8 {
+		input = append(input, make([]byte, peekSize-inputSize)...)
+	}
+
+	magic16 := binary.BigEndian.Uint16(input)
+	magic32 := binary.BigEndian.Uint32(input)
+	// We need to pre-pend the padding since we're reading into something bigendian and exceeding the
+	// 48bits size of the magic number bytes. The 16 and 32 bit magic numbers are complete bytes and
+	// therefore do not need any padding.
+	magic48 := binary.BigEndian.Uint64(append(make([]byte, 2), input[0:6]...))
+
+	switch true {
+	case magic16 == gzipMagic:
+		log.Debug().
+			Str("type", "gzip").
+			Msg("Compression Format")
+		return gzipDecompressor{}
+	case magic16 == bzipMagic:
+		log.Debug().
+			Str("type", "bzip2").
+			Msg("Compression Format")
+		return bzip2Decompressor{}
+	case magic16 == lzwMagic:
+		compressionByte := input[2]
+		// litWidth is guaranteed to be at least 9 per specification, the high order 3 bits of byte[2] are the litWidth
+		// the low order 5 bits are only used by non-unix implementations, we are going to ignore them.
+		litWidth := int(compressionByte>>5) + 9
+		log.Debug().
+			Str("type", "lzw").
+			Int("litWidth", litWidth).
+			Msg("Compression Format")
+		return lzwDecompressor{
+			order:    lzw.MSB,
```

Comment:
lzw.MSB - is it true for all streams? i think calculation of litWidth affected by it.